### PR TITLE
Object GC for pipelining

### PR DESCRIPTION
**Pull request by jianoaix**
## Why are these changes needed?
We leak memory when we create a DatasetPipeline from a "collapsed" DatasetPipeline (which executes multiple stages and produces a stream of output Datasets as the input base Datasets for the new DatasetPipeline).
The DatasetPipeline splitting is such a collapsing operation, so the child pipelines will have zero stages (no matter how many stages the parent pipeline had), which will make us no longer able to tell if it's safe to clear output blocks of child pipelines.

This PR fixes this by preserving whether the base Datasets can be cleared when we create new DatasetPipeline from the old.

The following script will spill and then OOM before; with this PR it should be able to run indefinitely with about 5GiB memory.

Next:
- Handle the copies created by block splitting when making equal pipeline splitting

```
import ray

dataset = ray.data.range_tensor(500000, shape=(80, 80, 4), parallelism=100)
p = dataset.window(bytes_per_window=1024 * 1024 * 1024).repeat()
p1, p2 = p.split(2, equal=True)
print("start")

@ray.remote
def f(p):
        i = 0
        for batch in p.iter_batches():
                i += 1
                print(i)

ray.get([f.remote(p1), f.remote(p2)])
```

## Related issue number

#25249

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — APPROVED** [review]

File: python/ray/data/dataset_pipeline.py
Comment (ericl) [inline python/ray/data/dataset_pipeline.py:86]: Can we actually make this a required variable? Otherwise it's easy to forget to set this properly.

**Review by clarkzinzow — APPROVED** [review]
Looks good to merge after fixing the merge conflicts!

**Pull request merged**
